examples/examples_test.sh

```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+set +e
+
+clean () {
+    jobs -p | xargs pkill -P 
+    wait
+}
+
+fail () {
+    echo "$(tput setaf 1) $1 $(tput sgr 0)"
+    clean
+    exit 1
+}
+
+pass () {
+    echo "$(tput setaf 2) $1 $(tput sgr 0)"
+}
+
+EXAMPLES=(
+    "helloworld"
+    "route_guide"
+    "features/authentication"
+    "features/compression"
+    "features/deadline"
+    "features/encryption/TLS"
+    )
+
+declare -A EXPECTED_SERVER_OUTPUT=( 
+    ["helloworld"]="Received: world"
+    ["route_guide"]=""
+    ["features/authentication"]="server starting on port 50051..."
+    ["features/compression"]="UnaryEcho called with message \"compress\""
+    ["features/deadline"]=""
+    ["features/encryption/TLS"]=""
+    )
+
+declare -A EXPECTED_CLIENT_OUTPUT=(
+    ["helloworld"]="Greeting: Hello world"
+    ["route_guide"]="location:<latitude:416851321 longitude:-742674555 >"
+    ["features/authentication"]="UnaryEcho:  hello world"
+    ["features/compression"]="UnaryEcho call returned \"compress\", <nil>" 
+    ["features/deadline"]="wanted = DeadlineExceeded, got = DeadlineExceeded" 
+    ["features/encryption/TLS"]="UnaryEcho:  hello world"
+    )
+
+for example in ${EXAMPLES[@]}; do
+    echo "$(tput setaf 4) testing: ${example} $(tput sgr 0)"
+
+    # Build server
+    if ! go build -o /dev/null ./examples/${example}/*server/*.go; then
+        fail "failed to build server"
+    else
+        pass "successfully built server"
+    fi
+
+    # Build client
+    if ! go build -o /dev/null ./examples/${example}/*client/*.go; then
+        fail "failed to build client"
+    else
+        pass "successfully built client"
+    fi
+    
+    # Start server
+    SERVER_LOG="$(mktemp)"
+    go run ./examples/$example/*server/*.go &> $SERVER_LOG  &
+
+    CLIENT_LOG="$(mktemp)"
+    if ! go run examples/${example}/*client/*.go &> $CLIENT_LOG; then
```

Comment:
Wouldn't this potentially introduce flakiness? If the server hasn't started serving before the client attempts a connection, this would fail even though it shouldn't.

Ideally, we'd want to start the server and wait for a string (like "serving on port whatever") to be printed before starting the client? I don't know if all examples have such a string. Something like `tail -f ${SERVER_LOG} | grep -q "some string"`?

Reply:
Yes this does produce flakiness. This is fixed by adding a `WithBlock` dial option to the client which was decided in this PR: 

https://github.com/grpc/grpc-go/pull/3092

I like the idea of waiting for the server to output to the server log more. However not all example servers output a string. I could add a string like "listening on port :50051" to every server example that doesn't have it. 

Would it be better to do this method and not use the `WithBlock` dialOption method?

Reply:
The clients have been updated to do a blocking dial (which will retry the dialing until a connection is established), so we should be good here.

Reply:
my bad, `WithBlock` is a better idea than the `tail | grep` thing. I didn't notice it, sorry!